python/ray/data/_internal/block_batching/iter_batches.py

```diff
@@ -0,0 +1,311 @@
+                num_rows_in_last_block = 0
+                global_index += 1
+
+            if buffer_size > batch_size:
+                # If current buffer is greater than batch size, then yield part of the
+                # buffer, and carryover the remainder to the next batch.
+                num_rows_to_leave_behind = buffer_size - batch_size
+                ending_index = num_rows_in_last_block - num_rows_to_leave_behind
+                assert ending_index > 0, ending_index
+                yield LogicalBatch(
+                    global_index, batch_buffer, starting_index, ending_index, batch_size
+                )
+                global_index += 1
+                # Carryover to next batch.
+                batch_buffer = [batch_buffer[-1]]
+                buffer_size = num_rows_to_leave_behind
+                starting_index = ending_index
+
+    # Yield any leftover batches if necessary.
+    if buffer_size > 0 and not drop_last:
+        assert buffer_size < batch_size
+        yield LogicalBatch(
+            global_index, batch_buffer, starting_index, None, buffer_size
+        )
+        global_index += 1
+
+
+def local_shuffle_logical_batches(
+    logical_batch_iterator: Iterator[LogicalBatch],
+    shuffle_buffer_min_size: int,
+    shuffle_seed: Optional[int] = None,
+) -> Iterator[LogicalBatch]:
+    """Shuffles the logical batch iterator using a buffer of the provided size."""
+
+    if shuffle_seed is not None:
+        random.seed(shuffle_seed)
+
+    shuffle_buffer: List[LogicalBatch] = []
+    shuffle_buffer_size = 0
+    global_counter = 0
+
+    for logical_batch in logical_batch_iterator:
+        shuffle_buffer.append(logical_batch)
+        shuffle_buffer_size += logical_batch.num_rows
+
+        while shuffle_buffer_size >= shuffle_buffer_min_size:
+            output_batch = shuffle_buffer.pop(
+                random.randint(0, len(shuffle_buffer) - 1)
+            )
+            output_batch.batch_idx = global_counter
+            yield output_batch
+            shuffle_buffer_size -= output_batch.num_rows
+            global_counter += 1
+
+    # Yield any leftover.
+    while len(shuffle_buffer) > 0:
+        output_batch = shuffle_buffer.pop(random.randint(0, len(shuffle_buffer) - 1))
+        output_batch.batch_idx = global_counter
+        yield output_batch
+        global_counter += 1
+
+
+def prefetch_batches_locally(
+    logical_batch_iter: Iterator[LogicalBatch],
+    prefetcher: BlockPrefetcher,
+    num_batches_to_prefetch: int,
+) -> Iterator[LogicalBatch]:
+    """Given an iterator of logical batches, returns an iterator over the same logical
+    batches, while prefetching `num_batches_to_prefetch` batches in advance.
+
+    Args:
+        logical_batch_iter: An iterator over logical batches.
+        prefetcher: The prefetcher to use.
+        num_batches_to_prefetch: The number of batches to prefetch ahead of the
+            current batch during the scan.
+    """
```

Comment:
The logic for this one seems a bit complicated to me, would something like this suffice?

```
buffer = []

for batch in logical_batch_iter:
   prefetch_blocks_of(batch)
   buffer.append(batch)
   if len(buffer) >= num_batches_to_prefetch:
       yield buffer.pop(0)

for b in buffer:
   yield b
```

Reply:
+1

Reply:
Updated the logic to be more similar to the previous logic with a sliding window now that we've decided to do block resolution in a single thread.